pkg/bindings/connection.go

```diff
@@ -17,46 +25,170 @@ type APIResponse struct {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
 	}
-	newConn := Connection{
-		url:    url,
-		client: &http.Client{},
+	// TODO once ssh is implemented, remove this block and
+	// add it to the conditional beneath it
+	if u.Scheme == "ssh" {
+		return nil, ErrNotImplemented
+	}
+	if u.Scheme != "tcp" && u.Scheme != "unix" {
+		return nil, errors.Errorf("%s is not a support schema", u.Scheme)
+	}
+
+	if u.Scheme == "tcp" && !strings.HasPrefix(uri, "tcp://") {
+		return nil, errors.New("tcp URIs should begin with tcp://")
+	}
+
+	address := u.Path
+	if u.Scheme == "tcp" {
+		address = u.Host
+	}
+	newConn := newConnection(u.Scheme, address)
+	ctx := context.WithValue(context.Background(), "conn", &newConn)
+	if err := pingNewConnection(ctx); err != nil {
+		return nil, err
 	}
-	response, err := http.Get(pingConnection)
+	return ctx, nil
+}
+
+// pingNewConnection pings to make sure the RESTFUL service is up
+// and running. it should only be used where initializing a connection
+func pingNewConnection(ctx context.Context) error {
+	conn, err := GetConnectionFromContext(ctx)
 	if err != nil {
-		return newConn, err
+		return err
+	}
+	// the ping endpoint sits at / in this case
+	response, err := conn.DoRequest(nil, http.MethodGet, "../../../_ping", nil)
+	if err != nil {
+		return err
+	}
+	if response.StatusCode == http.StatusOK {
+		return nil
 	}
-	if err := response.Body.Close(); err != nil {
-		return newConn, err
+	return errors.Errorf("ping response was %q", response.StatusCode)
+}
+
+// newConnection takes a scheme and address and creates a connection from it
+func newConnection(scheme, address string) Connection {
+	client := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial(scheme, address)
+			},
+		},
 	}
-	return newConn, err
+	newConn := Connection{
+		client:  &client,
+		address: address,
+		scheme:  scheme,
+	}
+	return newConn
 }
 
-func (c Connection) makeEndpoint(u string) string {
-	return fmt.Sprintf("%s%s", defaultConnection, u)
+func (c *Connection) makeEndpoint(u string) string {
+	// The d character in the url is discarded and is meaningless
+	return fmt.Sprintf("http://d/%s%s", defaultConnectionPath, u)
 }
 
-func (c Connection) newRequest(httpMethod, endpoint string, httpBody io.Reader, params map[string]string) (*APIResponse, error) {
-	e := c.makeEndpoint(endpoint)
+// DoRequest assembles the http request and returns the response
+func (c *Connection) DoRequest(httpBody io.Reader, httpMethod, endpoint string, queryParams map[string]string, pathValues ...string) (*APIResponse, error) {
+	var (
+		err            error
+		response       *http.Response
+		safePathValues []string
+	)
+	// Make sure path values are http url safe
+	for _, pv := range pathValues {
+		safePathValues = append(safePathValues, url.QueryEscape(pv))
+	}
+	safeEndpoint := fmt.Sprintf(endpoint, safePathValues)
```

Comment:
Shouldn't safePathValues be safePathValues[:] ?

Reply:
@jwhonce the linter puked on that.